### ISSUE TITLE
[Feature]: Model Generation Monitoring and Intervention (including 1. request-level thinking budget control as supported in Qwen/Claude APIs 2. repetition detection and truncation)

### ISSUE THREAD

**Issue by YouCanLX** [vllm-project/vllm#29791]
### Summary
We are proposing two model generation monitoring and intervention features for more intelligent, efficient, controlled generation:

- [ ] Request-level thinking budget control as supported in Qwen/Claude APIs

- [ ] Efficient and async repetition detection and truncation

These two features support most reasoning models, including open-source openai gpt-oss, deepseek, qwen, and doubao, and all non-reasoning models in a plug-and-play way, as the  generation monitoring and intervention are model agnostic and only modifying the final sampled tokens with guided decoding. 

No fine-tuning is needed! 

### Integration with Existing Techniques
We tested on a variant of vllm, and these two features successfully support simple integration with all the below features:
- Architecture-agnostic: dense/MoE  and full/sparse models are supported
- Complied CUDA graph
- PD disaggregation
- Multi-batch inference with continuous batch
- Multi-token speculative decoding including MTP, ngram, SAM decoding. (EAGLE is not tested)
- Online request-level thinking budget control
- Low overhead and async repetition detection and truncation for both thinking and summary stages

If this feature is aligned with vLLM's roadmap, we would be glad to share our code and submit a PR for review.😄 

### Detailed Description
**Thinking budget control**

To make LLM serving more intelligent and adaptive, we propose a system that automatically optimizes reasoning effort based on query semantics.

- Current Gap: Thinking models generate internal reasoning tokens without user-controllable limits, leading to potential inefficiency.

- Proposed Solution:
  - Core Feature: Expose a request-level thinking_budget parameter (as seen in Qwen/Claude) for manual user control.
  - Future Advanced Automation: Integrate this parameter with the vLLM semantic router. The flow would be:
     - The vllm_semantic_router classifies the incoming query.
    - Based on the classification (e.g., "mathematical," "summarization," "coding"), a predefined thinking_budget is automatically appended to the generation config.
    - vLLM executes the request with this optimized, context-aware budget.

This creates a closed-loop system that dynamically allocates reasoning compute, improving cost-effectiveness and latency for end-users.

**Repetition detection and intervention**

- Problem: Under certain conditions, models (both with and without chain-of-thought reasoning) can fall into a state of generating endless, repetitive content. This is more severe and frequent when using quantized or sparse model versions.
- Impact: Users must manually halt the generation, resulting in a poor experience and the loss of all intermediate work/tokens, without receiving a usable final answer.
- Request: Implement a fix or a generation-time safeguard to detect and break repetition loops.

### Related Issues and PRs
[Qwen3 repetition bug](https://github.com/vllm-project/vllm/issues/25977)

[Previous issue of vllm repetition generation with model quantization](https://github.com/vllm-project/vllm/issues/5663)

There is [an open PR for limit thinking tokens](https://github.com/vllm-project/vllm/pull/20859), but we are proposing a more unified and systematic solution.

### What's more
More CoT monitoring features, including overthinking and underthinking detection and truncation with a SLM verifier, can be implemented for safer, more intelligent, and controllable model generation [1, 2].

Apart from reasoning models, the non-thinking models will also benefit from these features. Therefore, we believe that they are general infrastructure enhancement for vllm. 

_**To enable next-generation features like CoT monitoring, adaptive reasoning budgets and their semantic-router-driven optimization, vLLM's engine requires new native capabilities.**_

[1] Korbak, Tomek, Mikita Balesni, Elizabeth Barnes, Yoshua Bengio, Joe Benton, Joseph Bloom, Mark Chen et al. "Chain of thought monitorability: A new and fragile opportunity for ai safety." arXiv preprint arXiv:2507.11473 (2025).
[2] Lin, Weizhe, Xing Li, Zhiyuan Yang, Xiaojin Fu, Hui-Ling Zhen, Yaoyuan Wang, Xianzhi Yu, Wulong Liu, Xiaosong Li, and Mingxuan Yuan. "TrimR: Verifier-based Training-Free Thinking Compression for Efficient Test-Time Scaling." arXiv preprint arXiv:2505.17155 (2025).

Since the key modifications involve the post-processing sampling, we would be grateful for your support. @DarkLight1337 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions
